### PR TITLE
Polymorphic queries -> discriminated union types

### LINKED ISSUE #454: Better return types for `.__type__.name` and polymorphic queries if possible
**Describe the proposed feature**
A clear and concise description of the syntax, what you want to happen, and why it's needed.

### Part 1. `__type__.name`
Currently the return type of `.__type__.name` is just `string`. We know the name of the object type that's being selected, so it should be possible to infer this as the correct string literal (or union of string literals if it's a parent type).

### Part 2. Polymorphic queries
Currently a query like this:
```
e.select(e.Content, () => ({
  title: true,
  ...e.is(e.Movie, { release_year: true }),
  ...e.is(e.TVShow, { num_seasons: true }),
}));
```
gets a return type like:
```
{
  title: string;
  release_year: number | null;
  num_seasons: number | null;
}[]
```
While this is technically correct, it would be more useful if we could return a discriminated union. I think it might be possible for the querybuilder to automatically insert the typename in the shape, and return a type like so:
```
({
  title: string;
} & ({
  __typename: 'default::Movie';
  release_year: number;
} | {
  __typename: 'default::Show';
  num_seasons: number;
}))[]
```

### PR DESCRIPTION
**Pull request by CarsonF**
I revived #508

Fixes #454

Probably a breaking change since the types are getting stricter.

### CODE REVIEW

**Comment by CarsonF** [comment]
I restored some of the old functionality, and gated the stricter types behind an opt-in via declaration merging. Just looking for ways to get this out without a major. But open to reverting it or other suggestions.

File: packages/generate/src/edgeql-js/generateObjectTypes.ts
Comment (CarsonF) [inline packages/generate/src/edgeql-js/generateObjectTypes.ts:188]: This doesn't appear to handle grand-children/parents.

**Comment by CarsonF** [comment]
Some scenario in my codebase is producing `{ __typename: string, ... } |` which breaks the narrowing with the union-ed concrete types. Digging...

**Comment by CarsonF** [comment]
Even though the implicit computed `__typename` produce the right string literals, these cannot yet be explicitly referenced.
```ts
e.shape(e.X, scope => ({
  __typename: scope.__type__.name
})
```
still produces a `string` not `"default::X"`

File: packages/generate/src/syntax/path.ts
Comment (CarsonF) [inline packages/generate/src/syntax/path.ts:52]: I'm not sure I understand the difference between Scalar's `TsType` & `TsArgType`. I was thinking that they were the output / input types respectively.

I don't understand why the input type would be called out here explicitly and set to the literal string - where/how would that be used be as an input?

I figured the output type remained as a `string` for BC reasons. So I created another future option to make it strict.
This is for this btw:
```ts
scope => ({ t: scope.__type__.name })
```

But then I was thinking about this case
```ts
scope => ({ __type__: { name: true } })
```
Which had already made the output type a strict string.

So I thought the other case was a miss, but there is some explicit code here.

I guess all that to say I could just do this:
```suggestion
      ? ScalarType<"std::str", Root[typeof typenameSymbol]>
```

File: integration-tests/lts/select.test.ts
Comment (CarsonF) [inline integration-tests/lts/select.test.ts:1146]: This PR doesn't yet create a discriminated union from this 😢

Comment (CarsonF) [reply integration-tests/lts/select.test.ts:1146]: This is result is the same in master. Since it hasn't regressed in this PR, I'll leave it for a separate PR.

File: packages/generate/src/syntax/toEdgeQL.ts
Comment (CarsonF) [inline packages/generate/src/syntax/toEdgeQL.ts:1403]: The new functionality implies a `__typename` field when needed. If the user specifies this and it doesn't match what is assumed then the types will be wrong. We could check the `__typename` value here if it is already present (from a user) and confirm it is `.__type__.name` or throw.
I couldn't figure out a clean way to do this because the actual generated EdgeQL explicitly references the scope, which I couldn't figure out how to access.
```edgeql
single __typename := __scope_1_defaultMovie.__type__.name,
```

**Review by scotttrinh — APPROVED** [review]
Very solid 👏 Thanks for the huge effort here. Love that we now have some feature-flagging/opt-in machinery set up in the typesystem, too, that seems generally very useful.

**Comment by scotttrinh** [comment]
One thing I want to think about is maybe having some tests with the legacy behavior just to catch any regressions we make between now and when we switch the new behavior to the default in a future major release. I can add those.

File: packages/generate/src/syntax/toEdgeQL.ts
Comment (scotttrinh) [inline packages/generate/src/syntax/toEdgeQL.ts:1304]: Ah, we don't actually have a runtime representation of the type-level "opt in" for this behavior, so this ends up still being an arguably breaking change since we will be returning more. Definitely arguable as to whether or not returning strictly more data is breaking, but anyone who currently has tests for the existing behavior will get failures.

I wonder if we can make a runtime configuration, and actually _infer_ the type-level config from that runtime config?

Comment (CarsonF) [reply packages/generate/src/syntax/toEdgeQL.ts:1304]: I probably wouldn't say that adding an additional field is a breaking change. User tests may not break. Only if they are asserting exactly or using snapshots.

But I can see how it's not black/white. IMO none of this stuff is. kinda like TS changes lol

File: packages/generate/src/syntax/path.ts
Comment (CarsonF) [reply packages/generate/src/syntax/path.ts:52]: @scotttrinh thoughts here? I don't think we need a flag for this & just use my suggestion

File: packages/generate/src/syntax/toEdgeQL.ts
Comment (scotttrinh) [reply packages/generate/src/syntax/toEdgeQL.ts:1304]: > User tests may not break. Only if they are asserting exactly or using snapshots.

I think this is _pretty_ common: make a query (or use a data layer more likely) and assert that the returned data is precisely these objects. I agree in general that if you are introducing new properties/data in your own application, maybe you don't consider that a breaking change, but this might warrant a little more care to avoid tripping up people when upgrading point releases.

Comment (CarsonF) [reply packages/generate/src/syntax/toEdgeQL.ts:1304]: Are you looking for a CLI flag for this as well?

Comment (scotttrinh) [reply packages/generate/src/syntax/toEdgeQL.ts:1304]: Yeah, I think we should add it, like `--enable-polymorphism-as-discriminated-unions` as silly as that sounds. This is non-trivial, but I think it's worth it to avoid going with a major version bump here. I'm honestly open to the major version bump, but that will absolutely delay getting this released.

If you think you have the bandwidth to add the CLI flag, let me know. Otherwise, I'll stick it in the queue and try to get to it (or maybe have @diksipav pick this up) ASAP.

Comment (CarsonF) [reply packages/generate/src/syntax/toEdgeQL.ts:1304]:
```
--future-polymorphism-as-discriminated-unions
```

File: packages/generate/src/syntax/path.ts
Comment (scotttrinh) [reply packages/generate/src/syntax/path.ts:52]: Good question. I'm not sure, but I'm fine with going with the suggested change for now and looking at what the fallout is.

File: packages/generate/src/syntax/toEdgeQL.ts
Comment (scotttrinh) [reply packages/generate/src/syntax/toEdgeQL.ts:1304]: yeah, that was my first thought, but then I thought maybe in the next major we could allow you to still opt-into the old behavior, but now I'm thinking there is no advantage to keeping the old behavior around in the next major.

Comment (CarsonF) [reply packages/generate/src/syntax/toEdgeQL.ts:1304]: Yeah I don't see a reason to keep the old behavior long term.

File: packages/generate/src/syntax/typesystem.ts
Comment (scotttrinh) [inline packages/generate/src/syntax/typesystem.ts:371]: So like,

```suggestion
                      (typeof future["strictTypeNames"]) extends true
```

etc

File: packages/generate/src/syntax/typesystem.ts
Comment (CarsonF) [reply packages/generate/src/syntax/typesystem.ts:41]: I think it depends on what the DX goal is.

If the goal is to just use the cli flag and expect that `future.ts` is statically overridden, then no we don't need `TypesystemOptions` or the declaration merging on `SetTypesystemOptions`.

If the goal is to allow this to be set dynamically. Then yes we still need it.
In that case, devs would set the futures to true.
```ts
future.polymorphismAsDiscriminatedUnions = true;
future.strictTypeNames = true;
```
And then denote to TS that they intend to do that with the declaration merge.

Comment (scotttrinh) [reply packages/generate/src/syntax/typesystem.ts:41]: I don't think it makes sense for this to be dynamic in any way at all. You're either opting into this future behavior and will refactor all of your code to match it, or you're not. I also care about making sure the runtime behavior and types are aligned, so allowing (or really I guess encouraging) both seems dangerous.

Comment (CarsonF) [reply packages/generate/src/syntax/typesystem.ts:41]: Then yeah we don't need any of this. Just base it all of the futures object like you suggested

File: packages/generate/src/cli.ts
Comment (CarsonF) [inline packages/generate/src/cli.ts:242]: I don't think this is sufficient. We need to be able to specify which futures you are opting into. I could be ok with future A and have updated my codebase to account for it. But then future B is added and I would be automatically opted into it as well which is a breaking change I wasn't ready for.

Comment (scotttrinh) [reply packages/generate/src/cli.ts:242]: Yeah, @diksipav and I talked a bit about this yesterday and here are my thoughts:

1. From the perspective of a general feature of the generator package, this `future` behavior is similar to TypeScript's own `strict` behavior, which if you opt into directly is breaking between versions. It stands for some set of more granular configuration values, so I think this _is_ useful in general.
2. From the perspective of this future behavior, I expect most (all?) folks to want to opt into both, but also would be fine exposing granular flags for each here, too. I don't want to add tests for each permutation since we plan on making this the _only_ behavior in a future breaking change, but that does mean there might be bugs or poorly understood side-effects of having one or the other enabled, but not both.

I don't think we need to block on this, though, and I'm fine with trying to land this as-is now. I'll leave it up to @diksipav to decide if she wants to address this in this PR specifically, or as a follow-up PR.

Comment (diksipav) [reply packages/generate/src/cli.ts:242]: hey lets merge this (once it's ready) and I'll create right away a follow up PR with individual future flags

Comment (diksipav) [reply packages/generate/src/cli.ts:242]: @scotttrinh I'll merge this and start working on the PR with individual flags, if u think we should "swap" tests will do it in the next PR

**Pull request merged**
